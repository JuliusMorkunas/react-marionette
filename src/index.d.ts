import * as React from 'react';
//import * as Marionette from 'backbone.marionette';

type MarionetteView = any;
type MarionetteViewClass = any;

interface OwnProps {
    view: MarionetteViewClass;
    viewOptions: any;
    onUpdateOptions?: (view: MarionetteView, viewOptions: any, nextViewOptions: any) => boolean;
    className?: string;
}

export class MarionetteComponent extends React.Component<OwnProps> {
}

export class ReactBehavior {
}

export class ReactView {
    constructor(options?: {
        attributes?: any[];
        className?: string;
        el?: any;
        tagName?: string;
        render: () => React.ReactNode,
        mountPoint: 'onShow'|'onRender'
    });
}
