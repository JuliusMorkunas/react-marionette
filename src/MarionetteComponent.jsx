/**
 * Developer: Stepan Burguchev
 * Date: 2/5/2017
 * Copyright: 2015-2017 ApprovalMax
 *       All Rights Reserved
 *
 * THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF ApprovalMax
 *       The copyright notice above does not evidence any
 *       actual or intended publication of such source code.
 */

import Marionette from 'backbone.marionette';
import React from 'react';
import PropTypes from 'prop-types';

class MarionetteComponent extends React.Component {
    constructor(props) {
        super(props);
        this._el = null;
        this._view = null;
        this._hostRegion = null;
        this._regionManager = null;
    }

    componentDidMount() {
        this._regionManager = new Marionette.View();
        this._hostRegion = this._regionManager.addRegion('hostRegion', {
            el: this._el
        });
        this._rebuildView();
    }

    componentWillReceiveProps(nextProps) {
        if (this.props.view !== nextProps.view) {
            throw new Error(
                '[MarionetteComponent] error: `props.view` cannot be changed after the initial render.'
            );
        }
    }

    componentDidUpdate(prevProps) {
        if (this.props.onUpdateOptions) {
            this.props.onUpdateOptions(this._view, prevProps.viewOptions, this.props.viewOptions);
        }
        if (this._view.shouldViewRebuild && this._view.shouldViewRebuild(this.props.viewOptions)) {
            this._rebuildView();
        }
    }

    componentWillUnmount() {
        this._regionManager.destroy();
    }

    getRegion() {
        return this._hostRegion;
    }

    _rebuildView() {
        if (!this.props.view) {
            return;
        }

        const View = this.props.view;
        this._view = new View(this.props.viewOptions);
        this._hostRegion.show(this._view);
    }

    render() {
        return <div className={this.props.className} ref={el => (this._el = el)} />;
    }
}

MarionetteComponent.propTypes = {
    view: PropTypes.func.isRequired,
    viewOptions: PropTypes.object,
    onUpdateOptions: PropTypes.func,
    className: PropTypes.string
};

export default MarionetteComponent;
