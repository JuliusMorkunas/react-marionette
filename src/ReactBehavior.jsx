/**
 * Developer: Stepan Burguchev
 * Date: 2/5/2017
 * Copyright: 2015-2017 ApprovalMax
 *       All Rights Reserved
 *
 * THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF ApprovalMax
 *       The copyright notice above does not evidence any
 *       actual or intended publication of such source code.
 */

import Marionette from 'backbone.marionette';
import React from 'react';
import ReactDOM from 'react-dom';

export default Marionette.Behavior.extend({
    initialize(options) {
        if (!options.render) {
            throw new Error('Missing options: render');
        }
    },

    onRender() {
        if (this.options.mountPoint === 'onRender') {
            this._mountReactComponent();
        }
    },

    onShow() {
        if (this.options.mountPoint !== 'onRender') {
            this._mountReactComponent();
        }
    },

    onDestroy() {
        if (this._mounted) {
            ReactDOM.unmountComponentAtNode(this._getContainerEl());
        }
    },

    _mountReactComponent() {
        ReactDOM.render(
            this.options.render(),
            this._getContainerEl()
        );
        this._mounted = true;
    },

    _getContainerEl() {
        if (this.options.containerEl) {
            return this.$(this.options.containerEl)[0];
        }
        return this.el;
    }
});
