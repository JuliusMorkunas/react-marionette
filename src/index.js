export { default as MarionetteComponent } from './MarionetteComponent';
export { default as ReactBehavior } from './ReactBehavior';
export { default as ReactView } from './ReactView';
