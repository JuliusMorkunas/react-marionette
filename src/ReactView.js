/**
 * Developer: Stepan Burguchev
 * Date: 2/5/2017
 * Copyright: 2015-2017 ApprovalMax
 *       All Rights Reserved
 *
 * THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF ApprovalMax
 *       The copyright notice above does not evidence any
 *       actual or intended publication of such source code.
 */

import Marionette from 'backbone.marionette';
import ReactBehavior from './ReactBehavior';

export default Marionette.ItemView.extend({
    initialize(options) {},

    template: false,

    behaviors() {
        return {
            ReactBehavior: {
                behaviorClass: ReactBehavior,
                containerEl: null,
                reducer: this.options.reducer,
                store: this.options.store,
                render: this.options.render
            }
        };
    }
});
