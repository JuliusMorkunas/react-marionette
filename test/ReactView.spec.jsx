import React from 'react';
import Marionette from 'backbone.marionette';
import { mount } from 'enzyme';
import { ReactView } from '../src/index';

describe('ReactView', () => {
    it('should render react component', () => {
        const regionManager = new Marionette.RegionManager();
        const region = regionManager.addRegion('hostRegion', {
            el: document.createElement('div')
        });
        const view = new ReactView({
            className: 'react-wrapper',
            render: () => <div className="react-component">Hello, React</div>
        });
        region.show(view);
        const html = view.el.outerHTML;

        expect(html).toMatchSnapshot();
    });

    it('should render react component on render', () => {
        const view = new ReactView({
            className: 'react-wrapper',
            render: () => <div className="react-component">Hello, React</div>,
            mountPoint: 'onRender'
        });
        view.render();
        const html = view.el.outerHTML;

        expect(html).toMatchSnapshot();
    });
});
