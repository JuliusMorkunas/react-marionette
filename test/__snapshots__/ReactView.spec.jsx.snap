// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactView should render react component 1`] = `"<div class=\\"react-wrapper\\"><div class=\\"react-component\\">Hello, React</div></div>"`;

exports[`ReactView should render react component on render 1`] = `"<div class=\\"react-wrapper\\"><div class=\\"react-component\\">Hello, React</div></div>"`;
